### PR TITLE
ROS SITL: offboard setpoints

### LINKED ISSUE #1745: multiplatform/ros sitl: add mavros interface

### PR DESCRIPTION
**Pull request by thomasgubler**
# What's new
  * mavlink dummy node which forwards a small subset of messages via udp to a mavros node
  * Support for local position and attittude offboard setpoints (XXX_target mavlink messages) in ROS SITL
  * Small demo node for offboard position control
  * launch files which launch the demo node and setup mavros: `roslaunch px4 gazebo_ardrone_empty_world_offboard_demo.launch`

Here is a small overview:
```
(Gazebo <--> ) [demo_positon_setpoint_generator node] <--> [mavros] <--> UDP <--> [mavlink node] <--> [controller nodes] <--> Gazebo
```

If you use your own (pizza delivery, obstacle avoidance) ros node instead of `[demo_positon_setpoint_generator node]` this will allow you to test the functionality in gazebo before replacing the part right of `UDP` with the real system (which is then connected to mavros through a serial port instead of UDP).

# Testing status
  * offboard position control is operational and tested with the demo node
  * offboard attitude control is operational and tested with the demo node

Includes #1786 

Fixes #1745

### CODE REVIEW

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -619,62 +578,53 @@ MavlinkReceiver::handle_message_set_position_target_local_ned(mavlink_message_t
...
-							!offboard_control_sp_ignore_position_some(offboard_control_sp)) {
+					/* set the local pos values */
+					if (!offboard_control_mode.ignore_position) {
 					pos_sp_triplet.current.position_valid = true;
-					pos_sp_triplet.current.x = offboard_control_sp.position[0];
-					pos_sp_triplet.current.y = offboard_control_sp.position[1];
-					pos_sp_triplet.current.z = offboard_control_sp.position[2];
+					pos_sp_triplet.current.x = set_position_target_local_ned.x;
+					pos_sp_triplet.current.y = set_position_target_local_ned.y;
+					pos_sp_triplet.current.z = set_position_target_local_ned.z;
```
Comment (bansiesta) [inline src/modules/mavlink/mavlink_receiver.cpp]: indentation?

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2014 PX4 Development Team. All rights reserved.
```
Comment (bansiesta) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: Year...

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2014 PX4 Development Team. All rights reserved.
```
Comment (bansiesta) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.h]: year...

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 4]
```diff
@@ -743,42 +693,29 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
...
-		offboard_control_sp.attitude_rate[0] = set_attitude_target.body_roll_rate;
-		offboard_control_sp.attitude_rate[1] = set_attitude_target.body_pitch_rate;
-		offboard_control_sp.attitude_rate[2] = set_attitude_target.body_yaw_rate;
 
 		/* set correct ignore flags for body rate fields: copy from mavlink message */
-		for (int i = 0; i < 3; i++) {
-			offboard_control_sp.ignore &=  ~(1 << (i + OFB_IGN_BIT_BODYRATE_X));
-			offboard_control_sp.ignore |=  (set_attitude_target.type_mask & (1 << i)) << OFB_IGN_BIT_BODYRATE_X;
-		}
+		offboard_control_mode.ignore_bodyrate = (bool)(set_attitude_target.type_mask & 0x7);
```
Comment (bansiesta) [inline src/modules/mavlink/mavlink_receiver.cpp]: Couldn't you add the nice bit defines?

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+	while (ros::ok()) {
+		loop_rate.sleep();
+		ros::spinOnce();
+
+		/* Publish example offboard position setpoint */
+		geometry_msgs::PoseStamped pose;
+		pose.pose.position.x = 0;
+		pose.pose.position.y = 0;
+		pose.pose.position.z = 1;
```
Comment (bansiesta) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: is z down here?

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 1 above)
Comment (thomasgubler) [inline src/modules/mavlink/mavlink_receiver.cpp]: Tagging #1788 as this stems from there. This is now a clear drawback of the rebase workflow...

Comment (thomasgubler) [inline src/modules/mavlink/mavlink_receiver.cpp]: Fixed in in #1786, will show up here when I rebase....

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 4 above)
Comment (thomasgubler) [inline src/modules/mavlink/mavlink_receiver.cpp]: #1786
defines would be nice, but mavlink should provide them

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp (hunk 5 above)
Comment (thomasgubler) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: yes, this is the ros/gazebo standard (damn ground robots... ;-) )

Comment (bansiesta) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: Fair enough

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp (hunk 2 above)
Comment (thomasgubler) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: fixed thanks

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.h (hunk 3 above)
Comment (thomasgubler) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.h]: fixed thanks

File: src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp (hunk 5 above)
Comment (thomasgubler) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: actually z is up here :)

Comment (thomasgubler) [inline src/platforms/ros/nodes/demo_offboard_position_setpoints/demo_offboard_position_setpoints.cpp]: ...so the the sign is correct here

File: launch/mavros_sitl.launch [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<launch>
+	<!-- vim: set ft=xml noet : -->
+	<!-- example launch script for PX4 based FCU's -->
+
+	<group ns="px4_multicopter">
```
Comment (AndreasAntener) [inline launch/mavros_sitl.launch]: Do we want this in the px4_namespace? If I'm doing something offboard, wouldn't I talk to a mavros node located at root level?

**Comment by AndreasAntener** [comment]
Offboard control works for me already, very cool :)
In the attitude demo the setpoints seem to alternate thrust, between 0.10000000149, 0.5 and nan . I guess this relates to the problem you were discussing in #1786

**Comment by LorenzMeier** [comment]
@thomasgubler do you have a specific recommendation for the resolution of it, or merely, do you see a chance to wrap within the next days?

File: launch/mavros_sitl.launch (hunk 6 above)
Comment (thomasgubler) [inline launch/mavros_sitl.launch]: good point, let's move it to root

**Comment by thomasgubler** [comment]
@AndreasAntener I never tested attitude. Did you run the demo which is work in progress or did you send your own setpoints?

**Comment by AndreasAntener** [comment]
I just ran the demo.

**Comment by AndreasAntener** [comment]
@thomasgubler I did some tests with automatically arming at this location: https://github.com/PX4/Firmware/pull/1793/files#diff-b722e8753a3ee0cd9578c391978844beR206

Somehow the controller interface plugin needs to create the dummy controller at startup, else we always have to fake manual input. I don't have enough insight already to propose something here..

Maybe we need some mavlink cmd handling for arming after all. If I have offboard logic that does the arming over mavros and checks the return result, it will break there. I guess we could just make the dummy mavlink node ACK every cmd, and the commander arms automatically as soon as offboard setpoints arrive. That way, pure integration tests that most likely don't arm beforehand will work, but also offboard logic manually executed. But it's more a convenience thing and not so important. In the end we want the commands to fail of course when we test things that rely on their execution..

**Comment by LorenzMeier** [comment]
The arming issue is the same for real flight. We need to support an off board / RC-less mode anyway. I think it would de-value the whole testing framework significantly if we don't try to make the test environment behave the same way as the real deal (even if it includes mocks).

**Comment by AndreasAntener** [comment]
I agree, for me it's more a question if this should be done in this or in the next iteration. If the next iteration would already mean porting the mavlink and commander stuff, creating a lot of mocks now is maybe not high prio. ok, at least the arming mock shouldn't be much ;)

**Comment by LorenzMeier** [comment]
A mock commander for arming sounds fair. We might need to C++ify the current commander into a class for the multi system support, and so having the mock would make everything less timing sensitive.

**Comment by thomasgubler** [comment]
> The arming issue is the same for real flight. We need to support an off board / RC-less mode anyway. I think it would de-value the whole testing framework significantly if we don't try to make the test environment behave the same way as the real deal (even if it includes mocks).

The problem is the lack of support for mavlink commands in the dummy commander (this node exists already) and in the dummy mavlink node. With the real system you can fly fine without RC by using the mavlink commands to switch to offboard and arm. So we can either duplicate the whole command handling or just use something that 'works' until we have the real port. I am in favor of the second.

**Comment by thomasgubler** [comment]
@AndreasAntener please check also https://github.com/PX4/euroc_simulator/tree/px4_nodes_upstreammerge3 the controller initialization will change after the next merge.

**Comment by thomasgubler** [comment]
The offboard attitude demo is now functional

**Comment by LorenzMeier** [comment]
Awesome! Looking forward to play with it!

**Comment by thomasgubler** [comment]
@AndreasAntener what is the testing level of your commits here? How are they related with #1831?
I rebased #1786 on master and this one one #1786 which means this branch now includes #1831 

From my point of view only #1768 is blocking this from merging

**Comment by LorenzMeier** [comment]
@thomasgubler I think you meant https://github.com/PX4/Firmware/pull/1786. I'm generally with the changes there, but I can't contribute to their testing - what would be needed to get those in?

**Comment by thomasgubler** [comment]
@LorenzMeier yes #1786 , in #1786 velocity control is not working. The most critical part (position control) was tested.

**Comment by AndreasAntener** [comment]
@thomasgubler my commits here are the result from the merge of #1831, so the todos from there apply. The mavros test currently fails, but that's actually what it should, since arming doesn't work yet. I still might make some minor updates to the tests soon, but if they go in now or after a merge to master doesn't matter.

**Comment by thomasgubler** [comment]
rebased on master an merging since #1786 is in

**Pull request merged**